aten/src/ATen/native/cpu/BlasKernel.cpp

```diff
@@ -112,29 +111,29 @@ void gemm_transa_(
 
 template <typename scalar_t, typename opmath_t>
 void gemm_transb_(
-    int64_t m, int64_t n, int64_t k,
+    int64_t m,
+    int64_t n,
+    int64_t k,
     opmath_t alpha,
-    const scalar_t *a, int64_t lda,
-    const scalar_t *b, int64_t ldb,
+    const scalar_t* a,
+    int64_t lda,
+    const scalar_t* b,
+    int64_t ldb,
     opmath_t beta,
-    scalar_t *c, int64_t ldc) {
-  // c *= beta
-  scale_(m, n, beta, c, ldc);
-
+    scalar_t* c,
+    int64_t ldc) {
   // c += alpha * (a @ b.T)
-  for (const auto l : c10::irange(k)) {
+  for (const auto i : c10::irange(m)) {
     for (const auto j : c10::irange(n)) {
-      opmath_t val = b[j + l * ldb] * alpha;
-      int64_t i_m = m / 4;
-      for (const auto i_i : c10::irange(i_m)) {
-        c[j * ldc + i_i * 4 + 0] += a[i_i * 4 + 0 + l * lda] * val;
-        c[j * ldc + i_i * 4 + 1] += a[i_i * 4 + 1 + l * lda] * val;
-        c[j * ldc + i_i * 4 + 2] += a[i_i * 4 + 2 + l * lda] * val;
-        c[j * ldc + i_i * 4 + 3] += a[i_i * 4 + 3 + l * lda] * val;
+      const auto dot = sum(k, [&](int64_t l) -> opmath_t {
+        return static_cast<opmath_t>(a[l * lda + i]) *
+            static_cast<opmath_t>(b[l * ldb + j]);
```

Comment:
The point of the old code is to access both arrays in contiguous order.

Reply:
Yes, it is. But I guess if I follow that, I will need malloc larger temporary `c` buffer to hold `accumulation value`, please kindly correct me if I am wrong.
Do you think it is necessary to keep this optimization in the `reference path` and might introduce larger memory usage?

Reply:
This is why I suggested specializing `gemm_core`. You can use the inefficient memory ordering for BFloat16 and Half, but the efficient ordering for all other types.

Reply:
I specializing `gemm_notrans_` and `gemm_transb_` to keep efficient ordering for all other types.